### PR TITLE
add support for custom tls fingerprint

### PR DESCRIPTION
**Pull request by VovkoO**
Hi, I added the ability to add a custom tls fingerprint. This feature would be very useful for me. I have added a new field to which you can add the hex client hello captured in wireshark.
Unfortunately, I had problems generating java code from .form file, I had to edit it manually.
Please take a look at my changes, maybe you will like them.

### CODE REVIEW

**Review by sleeyax — CHANGES_REQUESTED** [review]
A similar feature used to exist in V0 but I had to remove it in V1 due to time constraints, so thank you for adding it again! 

Left some initial comments, will do a more thorough review later.

File: src-go/server/certificate.go [hunk 1]
```diff
@@ -116,7 +118,7 @@ func NewCertificateAuthority() (*x509.Certificate, *rsa.PrivateKey, error) {
 		return nil, nil, err
 	}
 
-	err = os.WriteFile(caFile, raw, 0600)
+	err = os.WriteFile(caFile, raw, 0o600)
```
[the remark this answers is not in the document]
Comment (VovkoO) [reply src-go/server/certificate.go:121]: "0o" is just a way to write octal numbers, the permission doesnt change.

File: docs/screenshot.png
[the remark this answers is not in the document]
Comment (VovkoO) [reply docs/screenshot.png:1]: ok, fixed it

File: src-go/server/certificate.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sleeyax) [reply src-go/server/certificate.go:121]: I see. This project is currently formatted with regular `gofmt`. I don't think the usage of a stricter formatters like `gofumpt` hurt the code base though, so I'll accept the change.

File: docs/screenshot.png
Comment (sleeyax) [inline docs/screenshot.png:1]: Much better, but could you add an example hex client hello again? Just so people can see what this field would look like in practice.

File: src-go/server/go.mod [hunk 2]
```diff
@@ -4,6 +4,7 @@ go 1.20
 
 require (
 	github.com/ooni/oohttp v0.6.1
+	github.com/pkg/errors v0.9.1
```
Comment (sleeyax) [inline src-go/server/go.mod:7]: This package is [deprecated](https://github.com/pkg/errors) and probably unneeded. Please remove if possible.

File: src-go/server/internal/tls/factory.go [hunk 3]
```diff
@@ -31,6 +33,9 @@ func (f *FactoryWithClientHelloId) NewUTLSConn(conn net.Conn, config *tls.Config
 	if clientHelloID == nil {
 		clientHelloID = DefaultClientHelloID
 	}
+	if f.ClientHelloSpec != nil {
+		clientHelloID = &utls.HelloCustom
+	}
```
Comment (sleeyax) [inline src-go/server/internal/tls/factory.go:38]: The `ClientHelloSpec` field has priority over `clientHelloID`, so there's no need to check for `clientHelloID` if `ClientHelloSpec` is already set:

```suggestion
        if f.ClientHelloSpec != nil {
		clientHelloID = &utls.HelloCustom
	} else if clientHelloID == nil {
		clientHelloID = DefaultClientHelloID
	}
	
```

**Review by sleeyax — CHANGES_REQUESTED** [review]

File: src/main/java/burp/SettingsTab.form [hunk 4]
```diff
@@ -71,58 +71,76 @@
...
+        </properties>
+      </component>
+      <component id="56d1e" class="javax.swing.JTextField" binding="textFieldHexClientHello">
+        <constraints>
+          <grid row="6" column="0" row-span="1" col-span="2" vsize-policy="0" hsize-policy="6" anchor="8" fill="1" indent="0" use-parent-layout="false">
+            <preferred-size width="150" height="-1"/>
+          </grid>
+        </constraints>
+        <properties>
+          <toolTipText value="Custom client hello in hex stream."/>
```
Comment (sleeyax) [inline src/main/java/burp/SettingsTab.form:141]:
```suggestion
          <toolTipText value="Custom client hello as hex stream"/>
```

Or:

```suggestion
          <toolTipText value="Custom client hello in hex stream format"/>
```

File: docs/screenshot.png
Comment (VovkoO) [reply docs/screenshot.png:1]: fixed

File: src-go/server/go.mod (hunk 2 above)
Comment (VovkoO) [reply src-go/server/go.mod]: fixed

**Review by sleeyax — APPROVED** [review]

**Pull request merged**